platforms/nuttx/src/px4/stm/stm32f4/px4io_serial/px4io_serial.cpp

```diff
@@ -294,49 +291,54 @@ ArchPX4IOSerial::_bus_exchange(IOPacket *_packet)
 
 	/* wait for the transaction to complete - 64 bytes @ 1.5Mbps ~426µs */
 	int ret;
+	irqstate_t irqs = enter_critical_section();
 
 	for (;;) {
 		ret = sem_timedwait(&_completion_semaphore, &abstime);
 
 		if (ret == OK) {
 			/* check for DMA errors */
 			if (_rx_dma_status & DMA_STATUS_TEIF) {
-				// stream transfer error, ensure all DMA is also stopped before exiting early
+				/* stream transfer error, ensure all DMA is also stopped before exiting */
 				_abort_dma();
 				perf_count(_pc_dmaerrs);
 				ret = -EIO;
 				break;
-			}
 
-			/* check packet CRC - corrupt packet errors mean IO receive CRC error */
-			uint8_t crc = _current_packet->crc;
-			_current_packet->crc = 0;
+			} else {
+				/* succesful transaction (crc can still fail) */
+				break;
+			}
 
-			if ((crc != crc_packet(_current_packet)) || (PKT_CODE(*_current_packet) == PKT_CODE_CORRUPT)) {
+		} else {
+			if (errno == ETIMEDOUT) {
+				/* something has broken - clear out any partial DMA state and reconfigure */
 				_abort_dma();
-				perf_count(_pc_crcerrs);
-				ret = -EIO;
+				perf_count(_pc_timeouts);
+				perf_cancel(_pc_txns);		/* don't count this as a transaction */
 				break;
 			}
-
-			/* successful txn (may still be reporting an error) */
-			break;
-		}
-
-		if (errno == ETIMEDOUT) {
-			/* something has broken - clear out any partial DMA state and reconfigure */
-			_abort_dma();
-			perf_count(_pc_timeouts);
-			perf_cancel(_pc_txns);		/* don't count this as a transaction */
-			break;
 		}
 
-		/* we might? see this for EINTR */
-		syslog(LOG_ERR, "unexpected ret %d/%d\n", ret, errno);
+		/* Loop in case we are interrupted on sleep */
 	}
 
-	/* reset DMA status */
+	/* Make sure the DMA is properly stopped (set en bit, prevent further interrupts) */
```

Comment:
Can we get here with out these already being stopped? If so the logic is still not correct.

Reply:
Yes, yesterday I managed to get it into situation where it would then hang on the EN bit again in the next iteration (after ~20 minutes). This time, it was the TX dma - I believe that this could happen when the TX dma fails after all necessary bytes are transmitted, the error does not get picked up (since it does not have an interrupt handler), if the rx succeeds, we would not stop the tx dma.

Reply:
The HW will disable the EN on count == 0 or Error - so there may be another issue lurking.

Reply:
Unless this it the circular setting

Reply:
We don't have that for the tx dma though.

Reply:
Ah my bad I remembered it wrong, it is on the RX.

Reply:
Once the UART TXDMAEN (sic) is set. The Tx has to end in either a ERROR or done count ==0.

So if there a code path that is negating TXDMAEN ? That is the bad guy and he should do the dma_stop(TX) 

The order of op wound be clear TXDMAEN in U[S]ART then stop DMA

Reply:
Resolved now, the issue was another race that we found.